eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -353,19 +357,31 @@ validateCollateralContainsNonADA bal =
 -- > (_,i_f) := txvldt tx
 -- > ◇ ∉ { txrdmrs tx, i_f } ⇒ epochInfoSlotToUTCTime epochInfo systemTime i_f ≠ ◇
 validateOutsideForecast ::
-  HasField "vldt" (Core.TxBody era) ValidityInterval =>
+  ( HasField "vldt" (Core.TxBody era) ValidityInterval,
+    HasField "_protocolVersion" (Core.PParams era) ProtVer
+  ) =>
+  Core.PParams era ->
   EpochInfo (Either a) ->
+  -- | Current slot number
+  SlotNo ->
   SystemStart ->
+  -- | Stability window, in terms of number of slots
+  Word64 ->
   ValidatedTx era ->
   Test (UtxoPredicateFailure era)
-validateOutsideForecast ei sysSt tx =
+validateOutsideForecast pp ei slotNo sysSt sw tx =
   {-   (_,i_f) := txvldt tx   -}
   case getField @"vldt" (body tx) of
     ValidityInterval _ (SJust ifj)
       | not (nullRedeemers (txrdmrs' $ wits tx)) ->
-          -- ◇ ∉ { txrdmrs tx, i_f } ⇒
-          failureUnless (isRight (epochInfoSlotToUTCTime ei sysSt ifj)) $ OutsideForecast ifj
+          unless (allowOutsideForecastTTL pp) $
+            -- ◇ ∉ { txrdmrs tx, i_f } ⇒
+            if ifj < addSlots sw (succ slotNo)
```

Comment:
I don't think `slotNo` is correct here. Specifically, the sound forecast range doesn't begin at the current slot, it begins at the slot of the last applied block. As far as I know, "the slot of the last applied block" is not easily available inside `Test` etc. I have a local prototype that eg adds it to `Globals`, but it's a nuisance.

Reply:
That's good to know, thanks! It would definitely be a more invasive change to get the slot of last applied to this rule (and divergence from the spec, etc). Having it in `Globals` is a bit odd, but certainly easier for the ledger code.

Reply:
The `eraEnd` logic here is what we're trying to emulate, for reference. I still need to do some careful off-by-one checking.

https://github.com/input-output-hk/ouroboros-network/blob/d3eaedd4e4326fd3252b1be660898ef9c1bf5257/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State/Infra.hs#L199-L207

Reply:
This feels like something that _should_ be in the ledger rules (and spec), since it does have an effect on transaction validity. Putting something mutable like this in `Globals` seems wrong...